src/context/node.js

```diff
@@ -0,0 +1,354 @@
+    }
+    node[ASSIGNED_SLOT_PROP] = undefined;
+    forEachContained(node, (cn) => cn.discover());
+  }
+
+  /**
+   * Creates the context node and automatically starts the discovery process.
+   *
+   * @param {!Node} node
+   */
+  constructor(node) {
+    /** @const {!Node} */
+    this.node = node;
+
+    /**
+     * Whether this node is a root. The Document DOM nodes are automatically
+     * considered as roots. But other nodes can become roots as well
+     * (e.g. shadow roots) via `setIsRoot()` API.
+     *
+     * @package {boolean}
+     */
+    this.isRoot = node.nodeType == DOCUMENT_NODE;
+
+    /**
+     * The root context node. Always available for a DOM node connected to a
+     * root node after the discovery phase.
+     *
+     * @package {?ContextNode}
+     */
+    this.root = this.isRoot ? this : null;
+
+    /**
+     * Parent should be mostly meaningless to most API clients, because
+     * it's an async concept: a parent context node can can be insantiated at
+     * any time and it doesn't mean that this node has to change. This is
+     * why the API is declared as package-private. However, it needs to be
+     * unobfuscated to avoid cross-binary issues.
+     *
+     * @package {?ContextNode}
+     */
+    this.parent = null;
+
+    /**
+     * See `parent` description.
+     *
+     * @package {?Array<!ContextNode>}
+     */
+    this.children = null;
+
+    /** @private {boolean} */
+    this.parentOverriden_ = false;
+
+    /** @const @private {function()} */
+    this.scheduleDiscover_ = throttleTail(
+      this.discover_.bind(this),
+      setTimeout
+    );
+
+    this.discover();
+  }
+
+  /**
+   * Requests the discovery phase. Asynchronously finds the nearest parent for
+   * this node and its root. Roots and parents set directly via `setParent()`
+   * API are not discoverable.
+   */
+  discover() {
+    if (this.isDiscoverable()) {
+      this.scheduleDiscover_();
+    }
+  }
+
+  /**
+   * Sets (or unsets) the direct parent. If the parent is set, the node will no
+   * longer try to discover itself.
+   *
+   * @param {!ContextNode|!Node|null} parent
+   */
+  setParent(parent) {
+    const parentContext = !parent
+      ? null
+      : parent.nodeType
+      ? ContextNode.get(/** @type {!Node} */ (parent))
+      : /** @type {!ContextNode} */ (parent);
+    this.updateTree_(parentContext, /* parentOverriden */ parent != null);
+  }
+
+  /**
+   * Designates (or undesignates) the node as a root node. If the node is
+   * designated as a root, it will no longer discover itself.
+   *
+   * @param {boolean} isRoot
+   */
+  setIsRoot(isRoot) {
+    this.isRoot = isRoot;
+    const newRoot = isRoot ? this : this.parent ? this.parent.root : null;
+    this.updateRoot(newRoot);
+  }
+
+  /**
+   * @param {?ContextNode} root
+   * @protected Used cross-binary.
+   */
+  updateRoot(root) {
+    const oldRoot = this.root;
+    if (root != oldRoot) {
+      // The root has changed.
+      this.root = root;
```

Comment:
Is there any need to also update `isRoot` for these nodes? I know this is done prior to the call to this method in `setIsRoot`, but I'm not sure how else this method may be used. For example it looks like this isn't dealt with in `updateTree_`.

Reply:
Not in this case. This just says that a node now has a root, but it's not a root itself - it's only connected to a root.

Reply:
It wouldn't be helpful to assert `!this.isRoot`?

Reply:
That's a good idea. Will do.